File: 19-1.py
#!/usr/bin/env python

import re

rule_specifications, messages = open('19.input').read().split("\n\n")

rules = {}

for rule in rule_specifications.split("\n"):
    num, prod = rule.split(':')
    if '"' in prod:
        prods = prod.strip().replace('"', '')
    elif '|' in prod:
        l = [[int(r) for r in p.split()] for p in prod.split('|')]
        prods = {}
        for i in range(len(l)):
            prods[i] = l[i]
    else:
        prods = [ int(p) for p in prod.split()]

    rules[int(num)] = prods

def build_regexp(num):
    if isinstance(num, str):
        prod = num
    elif isinstance(num, int):
        prod = build_regexp(rules[num])
    elif isinstance(num, list):
        prod = "".join([build_regexp(r) for r in num])
    elif isinstance(num, dict):
        prod = '(?:' + "|".join([build_regexp(r) for r in num.values()]) + ')'

    if isinstance(num, int) and isinstance(prod, str):
        rules[num] = prod

    return prod

build_regexp(0)

regexp = re.compile(rules[0])

count = 0
for m in messages.split('\n'):
    if regexp.fullmatch(m):
        count += 1

print(count)

File: 19-1.rb
#!/usr/bin/env ruby

rule_specifications, messages = File.read('19.input').split("\n\n")

$rules = {}

rule_specifications.lines.each do |rule|
  num, prod = rule.split ':'
  prods =
    if prod.include? '"'
      prod.strip.gsub('"', '')
    elsif prod.include? '|'
      h = {}
      prod.split('|').map { |p| p.split.map &:to_i }.each_with_index { |p, i| h[i] = p }
      h
    else
      prod.split.map &:to_i
    end
  $rules[num.to_i] = prods
end

def build_regexp(num)
  prod =
    if num.is_a? String
      num
    elsif num.is_a? Integer
      build_regexp $rules[num]
    elsif num.is_a? Array
      num.map { |r| build_regexp r }.join
    elsif num.is_a? Hash
      '(?:' + num.values.map { |r| build_regexp r }.join('|') + ')'
    end
  if num.is_a? Integer and prod.is_a? String
    $rules[num] = prod
  end
  prod
end

build_regexp 0

re = Regexp.new('^' + $rules[0] + '$')

print messages.lines.map(&:strip).count { |m| re.match? m }, "\n"
